docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,185 @@
+
+```golang
+// GCIS stands for Global Candidate Index Set
+sort(GCIS, GCIS.Score, Descend_Order)
+recommend_set = GCIS[:N]
+remaining_set = GCIS[N:]
+min_cost = EvaluateWorkload(workload, recommend_set)
+
+for {
+    v_recommend_set, v_remaining_set = SwapElements(recommend_set, remaining_set, swap_size)
+    variant_cost = EvaluateWorkload(workload, v_recommend_set)
+    if variant_cost < min_cost {
+        min_cost = variant_cost
+        recommend_set = v_recommend_set
+        remaining_set = v_remaining_set
+    }
+    ......
+    // loop until specific rounds or timeout, break
+}
+```
+----------------------------------------------
+### A quick exmaple for single-column index recommendation
+
+**Workload**:
+
+```
+- Query1: SELECT COUNT(*) FROM t1 where a = 1 and b = 5
+- Query2: SELECT a FROM t1 ORDER BY c limit 10 
+- Query3: DELETE FROM t1 WHERE a = 1 AND d = 1
+```
+where index on column *c* has already existed. 
+
+- **Step1**: recommend indexes for single query. For each query, compile it and get the `original_cost` output by SQL optimizer. Then add all possible hypothetic indexes *a, b, d* to the schema, compile the same query again. After the compiling, `LSIs` can be extracted from output execution plan as well as the `virtual_cost`. Denote `reduced_cost` as *original_cost - virtual_cost*. Assume this step outputs:
+
+```
+Query 1--LSIs: {a, b}, reduced_cost: 4
+Query 2--LSIs: {c},    reduced_cost: 4
+Query 3--LSIs: {a, d}, reduced_cost: 3
+```
+- **Step2**: get the score list. Simply add up `LSIs`' `reduced_cost` of each query respectively and then sort the list descending by score. Assume this step outputs:
+
+```
+a: 4+3=7
+b: 4
+c: 4
+d: 3
+```
+
+- **Step3**: evaluate recommend index set. Assume the number of allowed indexes recommended by index advisor is 2. Thus we can get the initial recommend index set composed of the top 2 indexes that is `{a, b}`. Then we add `{a, b}` to the schema at the same time mute existed index `c`. Having done that, compile all queries in the workload and add up each query's cost. Assume the sum is *20* and then assign it variable `min_cost`. Assume the *swap_size* is 1, that is every time randomly exchange 1 element between the recommend set and remaining set. Here `{a, b}` is the recommend set, `{c,d}` is the remaining set. Assume *a* and *c* is chosen respectively in recommend set and the remaining set, that is, now the recommend set is `{c, b}` and the remaining set is `{a, d}`. Apply recommend set `{c, b}` to the workload and get a sum cost, say *18*. As the cost decreased from *20* to *18*, update the recommend set from `{a, b}` to `{c, b}`. Then try swapping 1 element between `{c, b}` and `{a, d}` and execute the re-evaluate process. Note that if the sum cost with `{c, b}` is *22* ( greater than 20 ) instead of *18*, the next *swap* process should happen between `{a, b}` and `{c, d}` instead of between `{c, b}` and `{a, d}`. Loop the *Swap and Re-evaluate* process until timeout.
```

Comment:
Really recommend to split it into smaller parts...

Reply:
OK~